pip/wheel.py

```diff
@@ -631,26 +657,126 @@ def supported(self, tags=None):
+            os.environ.pop('PATH', None)
+        else:
+            os.environ['PATH'] = self.save_path
+
+        if self.save_pythonpath is None:
+            os.environ.pop('PYTHONPATH', None)
+        else:
+            os.environ['PYTHONPATH'] = self.save_pythonpath
+
+        if not self.no_clean:
+            rmtree(self.prefix)
+
+        return False  # Do not suppress exceptions
+
+
 class WheelBuilder(object):
     """Build wheels from a RequirementSet."""
 
     def __init__(self, requirement_set, finder, build_options=None,
-                 global_options=None):
+                 global_options=None, no_clean=False):
         self.requirement_set = requirement_set
         self.finder = finder
         self._cache_root = requirement_set._wheel_cache._cache_dir
         self._wheel_dir = requirement_set.wheel_download_dir
         self.build_options = build_options or []
         self.global_options = global_options or []
+        self.no_clean = no_clean
+
+    def _find_build_reqs(self, req):
+        """Get a list of the packages required to build the project, if any.
+
+        Build requirements can be specified in a pyproject.toml, as described
+        in PEP 518
+        """
+        if os.path.isfile(req.pyproject_toml):
+            with open(req.pyproject_toml) as f:
+                pp_toml = pytoml.load(f)
+            return pp_toml.get('build-system', {}).get('requires', [])
+
+        return []  # No pyproject.toml
+
+    def _install_build_reqs(self, reqs, prefix):
+        # Local import to avoid circular import (wheel <-> req_install)
+        from pip.req.req_install import InstallRequirement
+        urls = [self.finder.find_requirement(InstallRequirement.from_line(r),
+                                             upgrade=False).url
+                for r in reqs]
+
+        if sys.version_info >= (2, 7):
+            mpip = 'pip'
+        else:
+            mpip = 'pip.__main__'  # Python 2.6 can't execute a package with -m
+        args = [sys.executable, '-m', mpip, 'install', '--prefix', prefix] \
+            + list(urls)
+        with open_spinner("Installing build dependencies") as spinner:
+            call_subprocess(args, show_stdout=False, spinner=spinner)
 
     def _build_one(self, req, output_dir, python_tag=None):
         """Build one wheel.
 
         :return: The filename of the built wheel, or None if the build failed.
         """
+        build_reqs = self._find_build_reqs(req)
+
+        if build_reqs:
+            # Install build deps into temporary prefix (PEP 518)
+            with BuildEnvironment(no_clean=self.no_clean) as prefix:
+                self._install_build_reqs(build_reqs, prefix)
+                return self._build_one_inside_env(req, output_dir,
+                                                  python_tag=python_tag,
+                                                  isolate=True)
+        else:
+            # Old style build, in the current environment
```

Comment:
I'd probably not call it old style, since it'd also be legtimately the style that would be used if someone had a completely self contained build system yea?